compiler/rustc_resolve/src/macros.rs

```diff
@@ -1076,10 +1077,22 @@ impl<'a> Resolver<'a> {
 
         if result.is_builtin {
             // The macro was marked with `#[rustc_builtin_macro]`.
-            if let Some(ext) = self.builtin_macros.remove(&item.ident.name) {
+            if let Some(builtin_macro) = self.builtin_macros.get_mut(&item.ident.name) {
+                use crate::BuiltinMacroState::*;
+
                 // The macro is a built-in, replace its expander function
                 // while still taking everything else from the source code.
-                result.kind = ext.kind;
+                // If we already loaded this builtin macro, give a better error message than 'no such builtin macro'.
+                match mem::replace(builtin_macro, AlreadySeen(item.span)) {
+                    NotYetSeen(ext) => result.kind = ext.kind,
+                    AlreadySeen(span) => {
+                        let msg = "attempted to define built-in macro more than once";
+                        self.session
+                            .struct_span_err_with_code(item.span, msg, error_code!(E0773))
```

Comment:
IIRC, there's a macro (`struct_span_err!`?) that wraps an error into `error_code!` automatically, but I'm not sure.

Reply:
That allocates a new string when it doesn't have to: https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_errors/diagnostic_builder.rs.html#439-447 but I can switch it if you prefer to be consistent with the rest of the code base.

Reply:
>That allocates a new string when it doesn't have to

During error reporting it doesn't matter much.